### PR TITLE
Basic structure

### PR DESCRIPTION
**Pull request by fesiib**
Check this out, guys

### CODE REVIEW

File: src/generator.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        self.rate_crossover = rate_crossover
+        self.rate_mutation = rate_mutation
+        self.max_iter = max_iter
+        self.wait_iter = wait_iter
+        self.dependency_mode = dependency_mode
+        self.test_type = test_type
+        self.validate_global_variables()
+
+    def generate_tests(self, source_code: SourceCode):
+        ipdg = IPDG(source_code)
```
Comment (Rainhunter13) [inline src/generator.py:61]: i named class PDG instead of IPDG as thought it is not interface anymore, just a class de

File: src/generator.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        ct = CoverageTable(source_code)
```
Comment (Rainhunter13) [inline src/generator.py:62]: i changed init of CoverageTable from source_code to pdg) it was little more convenient sorry

File: src/generator.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        random_seed.initial_population(self.seed_size, self.test_type)
+        ct.update(random_seed)        
+
+        dependency_mode = 0
+
+        #for Statistics
+        generations_cnt = 0
+
+        while True:
+            target: Predicate = ct.get_target_branch()
```
Comment (Rainhunter13) [inline src/generator.py:76]: I added pdg parameter for get_target_branch() as it was really needed for that

File: src/generator.py (hunk 1 above)
Comment (fesiib) [reply src/generator.py:61]: okay, easy to adjust

File: src/generator.py (hunk 2 above)
Comment (fesiib) [reply src/generator.py:62]: what is pdg? at least what is the type or format?

File: src/generator.py (hunk 3 above)
Comment (fesiib) [reply src/generator.py:76]: why do we need pdg for get_target_branch()? it should be solely tied to coveragetable isn't it?

Comment (Rainhunter13) [reply src/generator.py:76]: it depends on eases, which are calculated based on coverages and paths(of pdg)

File: src/generator.py (hunk 2 above)
Comment (Rainhunter13) [reply src/generator.py:62]: look at pdg.py please, it's just 2 trees with true/false edges. I am taking predicates for cov table from pdg

File: src/generator.py (hunk 3 above)
Comment (fesiib) [reply src/generator.py:76]: is it pdg or cfd?

Comment (fesiib) [reply src/generator.py:76]: so ultimately init of CoverageTable should include both pdg and source_code right?

Comment (Rainhunter13) [reply src/generator.py:76]: no, only pdg i guess

File: src/generator.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+                generations_cnt += 1
+                while next_population.get_size() < self.population_size:
+                    (parent1, parent2) = cur_population.tournament_selection()
+                    offspring1: TestCase = parent1.copy()
+                    offspring2: TestCase = parent2.copy()
+                    if random.random() <= self.rate_crossover:
+                        (offspring1, offspring2) = cur_population.crossover(
+                            parent1, parent2
+                        )
+                    if random.random() <= self.rate_mutation:
```
Comment (allbackoff) [inline src/generator.py:115]: for a non-uniform mutation, we need a way to reduce the mutation rate. example:
`rate_mutation -= lowering_rate` every time the function is called.
and if we do it here, we don't need the `mode` parameter

File: src/generator.py (hunk 2 above)
Comment (Rainhunter13) [reply src/generator.py:62]: FORGET ABOUT THAT, I FIXED MY PART

File: src/generator.py (hunk 3 above)
Comment (Rainhunter13) [reply src/generator.py:76]: FORGET ABOUT THAT, I FIXED MY PART

**Pull request merged**